toolchain/lowering/lowering_context.h

```diff
@@ -45,6 +45,21 @@ class LoweringContext {
     lowered_nodes_[node_id.index] = value;
   }
 
+  // Gets a callable's function.
+  auto GetLoweredCallable(SemanticsCallableId callable_id) -> llvm::Function* {
+    CARBON_CHECK(lowered_callables_[callable_id.index] != nullptr)
+        << callable_id;
+    return lowered_callables_[callable_id.index];
+  }
+
+  // Sets a callable's function.
+  auto SetLoweredCallable(SemanticsCallableId callable_id,
+                          llvm::Function* function) {
+    CARBON_CHECK(lowered_callables_[callable_id.index] == nullptr)
+        << callable_id;
+    lowered_callables_[callable_id.index] = function;
+  }
```

Comment:
I consider callables to be closer to strings or integers in the hierarchy, rather than nodes. That is, a callable does not have a 1:1 correspondence to a node:

1) Right now, a call knows the CallableId it's supposed to refer to, not a NodeId. What you're suggesting would require changing that (Trying to choose a canonical NodeId, maybe? Particularly relevant once we get into forward declarations.) It's not obvious to me _how_ that should change though, since it's just not how I'm tracking callables in semantics right now -- do you want me to investigate?

2) A NodeId might have a different value than the CallableId, for example with a lambda (I'm not sure how semantics for that would work).

Reply:
I should  note, _particularly_ important is the idea of how a call learns whether it got a definition later in the file. Nodes will find the forward declaration, and the problem is currently addressed through CallableId, so this would be a significant redesign of functions to make nodes work instead.

Reply:
Makes sense. Could you document some of this in the comments?

FWIW, if you end up exploring the other direction, i'd imagine making the first declaration always be the canonical one, and later ones canonicalize back to it. Not sure I understand the challenge with a call getting a definition later in the file.

Reply:
Adding to lowered_callables_:

 // Semantics treats callables as the
  // canonical form of a function, so lowering needs to do the same.

> FWIW, if you end up exploring the other direction, i'd imagine making the first declaration always be the canonical > one, and later ones canonicalize back to it. Not sure I understand the challenge with a call getting a definition later > in the file.

import library "A"; // Forward-declares a function.
import library "B"; // Forward-declares the same function.

Libraries "A" and "B" don't import each other. Which one provides the canonical declaration?

Reply:
> > FWIW, if you end up exploring the other direction, i'd imagine making the first declaration always be the canonical > one, and later ones canonicalize back to it. Not sure I understand the challenge with a call getting a definition later > in the file.
> 
> import library "A"; // Forward-declares a function. import library "B"; // Forward-declares the same function.
> 
> Libraries "A" and "B" don't import each other. Which one provides the canonical declaration? [Just make an arbitrary choice?]

Yeah, I'd just make an arbitrary choice.